### PR TITLE
get backup name from restore spec when handling restore DownloadRequests

### LINKED ISSUE #398: Incorrect pre-signed urls generated for named restores
If you do `ark restore create some-name --from-backup foo`, we end up generating the wrong pre-signed urls for `ark restore describe` and `ark restore logs`. We need to generate a url for (bucket, backup "directory", file), and the code assumes that all restores are named like `<backup name>-<timestamp>`. But for user-named restores, that assumption isn't necessarily valid. If you have a restore named `r1` for a backup `b1`, the url needs to be (bucket, b1, r1-logs.gz), for example, but it ends up being (bucket, r1, r1-logs.gz).

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <steve@heptio.com>

Fixes #398

### CODE REVIEW

**Comment by ncdc** [comment]
@skriss this works, or we could modify the `DownloadTarget` struct to have a field for the name of the backup. Modifying the struct seems like a lot less code, but it does mean that we'd have to resolve `Name` and `BackupName` - would we keep `Name` or change it to `RestoreName`? WDYT re approach?

**Comment by skriss** [comment]
Yeah, I thought about going that way as well - I couldn't think of any significant pros/cons between them so just went with the one that didn't change the API. If we go that approach, we'd presumably still want to fetch the restore's backup and automatically fill that in, so that code would just move from the controller into the `downloadrequest.Stream` func.

**Comment by ncdc** [comment]
I'd put it into the `ark restore <cmd>` code, and just fail the download request in the controller if the required field(s) are missing.

**Comment by ncdc** [comment]
If we go that route

File: pkg/controller/download_request_controller.go [hunk 1]
```diff
@@ -214,8 +219,24 @@ const signedURLTTL = 10 * time.Minute
...
-	var err error
-	update.Status.DownloadURL, err = c.backupService.CreateSignedURL(downloadRequest.Spec.Target, c.bucket, signedURLTTL)
+	var (
+		directory string
+		err       error
+	)
+
+	switch downloadRequest.Spec.Target.Kind {
+	case v1.DownloadTargetKindRestoreLog, v1.DownloadTargetKindRestoreResults:
+		restore, err := c.restoreLister.Restores(v1.DefaultNamespace).Get(downloadRequest.Spec.Target.Name)
```
Comment (ncdc) [inline pkg/controller/download_request_controller.go]: s/v1.DefaultNamespace/downloadRequest.Namespace/

File: pkg/controller/download_request_controller.go [hunk 2]
```diff
@@ -214,8 +219,24 @@ const signedURLTTL = 10 * time.Minute
...
+		if err != nil {
+			return errors.Wrap(err, "error getting Restore")
+		}
+
+		directory = restore.Spec.BackupName
+	default:
+		directory = downloadRequest.Spec.Target.Name
```
Comment (ncdc) [inline pkg/controller/download_request_controller.go]: This flow technically isn't used anywhere - do you think it's worth it?

**Comment by ncdc** [comment]
@skriss let's just go with this approach. I'd rather not have the messy `name` vs `restoreName` api change.

File: pkg/controller/download_request_controller.go (hunk 2 above)
Comment (skriss) [reply pkg/controller/download_request_controller.go]: The `default` case? That should catch anything backup-related (logs, tarball)

**Comment by ncdc** [comment]
It’s not used, fwiw

**Comment by skriss** [comment]
Ah, I see what you mean. I removed that case and added a comment to godoc about if/how `directory` is used within `BackupService`.

**Comment by ncdc** [comment]
LGTM

**Pull request merged**